### PR TITLE
Allow extenders to register their custom profile types in the ProfilesCache - #1387

### PR DESCRIPTION
**Pull request by phaumer**
## Proposed changes

This implements the requirements from https://github.com/zowe/vscode-extension-for-zowe/issues/1387

See the updates to the docs/README-Extending.md file for details.

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [X] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [X] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [X] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [X] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [X] I have tested it manually and there are no regressions found
- [X] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

**Comment by phaumer** [comment]
@AndrewTwydell could you give this a try?

File: docs/README-Extending.md [hunk 1]
```diff
@@ -168,6 +168,37 @@ The FTP Zowe Explorer extension provides examples for providing a data provider
...
+
+To support that the Zowe Explorer profiles cache that you can access as described above offers register method that allows adding a profile type to the cache that is not associated with any of the three APIs list above. Note, that the profile type must be valid Zowe CLI profile that is installed on the end user's home directory. The extension needs to therefore make sure it called the `initForZowe()` before trying to register a custom type.
+
+```typescript
+// Retrieve the Zowe Explorer API object from the currently running instance.
+// It must be at least Zowe Explorer 1.18.0 or newer or undefined will returned.
+const zoweExplorerApi = ZoweVsCodeExtension.getZoweExplorerApi("1.18.0");
+if (zoweExplorerApi) {
+  // Initialized the users ~/.zowe directory with the metadata for FTP profiles in case
```
Comment (AndrewTwydell) [inline docs/README-Extending.md:182]: Not sure it matters too much, but this documentation references both CICS and FTP things - I know it's only an example section but might be worth a look over??

**Review by AndrewTwydell — APPROVED** [review]
@phaumer just given it a try and it worked wonders! 

I have used the `initForZowe` method, and then `registerCustomProfilesType` followed by `reloadProfiles`, and I can now access the CICS profiles and all methods that come with the ProfilesCache! 

It also prevents the CICS profiles leaking into other trees because of the 'bodge' way we were registering the CICS APIs.

Thanks again for implementing this - it's the last blocking issue that we wanted to get past! 😃 🥇

File: docs/README-Extending.md (hunk 1 above)
Comment (lauren-li) [reply docs/README-Extending.md]: Good catch!

**Review by lauren-li — DISMISSED** [review]
This PR works well for me. All unit and expected integration tests are passing, and I was able to successfully follow the documentation and code sample to implement the custom profile types in my fork of the CICS extension.

I did make some minor copyedits in the Extender docs to try and clarify some pieces. @phaumer and @AndrewTwydell, if you could please take another look at those to be sure the updates look good to you, that would be great! Thank you!

Thanks @phaumer for adding in this nice capability and well-commented code sample!

**Comment by phaumer** [comment]
Thanks @lauren-li !

**Review by AndrewTwydell — APPROVED** [review]
LGTM 👍🏼 

Thanks @phaumer for adding the `registerCustomProfilesType` method! Everything is working fine for the CICS extension when using this change. 🎉

**Comment by crawr** [comment]
I've tested this branch in Theia locally and all tests passed. I can check why the pipeline is failing but I am ready to merge this as soon as the other checks pass.

**Comment by phaumer** [comment]
@crawr I tried re-running the tests again and they fail with network issues.

**Review by lauren-li — APPROVED** [review]
As a new `yarn audit` vulnerability popped up, I made commits to add `**/@npmcli/git` to `resolutions` in the top-level `package.json` file. The PR functionality is still working well in manual testing with my fork of the CICS extension. All unit tests are passing. My integration tests are still acting up on all branches, but the commits on this PR since its last approval shouldn't have affected them.

@JillieBeanSim is checking this as well, and will hopefully be able to verify the expected integration tests are still passing. Thanks Billie!

**Comment by lauren-li** [comment]
@AndrewTwydell Apologies for asking you to review this PR yet again, but would you be able to check that the functionality is still working well for you in your implementation of the CICS extension after the latest changes? If so, that would be great! Thanks in advance!

**Comment by lauren-li** [comment]
The results from @JillieBeanSim's run of the integration tests show that all expected integration tests are passing. (The `Enter pattern` test is handled in PR #1421.) Thanks Billie!:

ze-pr1419-intTests

**Review by AndrewTwydell — APPROVED** [review]
@lauren-li LGTM 👍🏼 

I've been using the `add-custom-profile-types` branch in development with the CICS extension with no problems. This includes the `getZoweExplorerApi`, `initForZowe`, `registerCustomProfilesType`, `reloadProfiles` and `getProfilesCache` methods (as well as everything the profilesCache offers).

Looking forward to this being available in 1.18 !!

**Pull request merged**